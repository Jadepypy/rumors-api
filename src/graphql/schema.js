import { GraphQLObjectType, GraphQLSchema } from 'graphql';

// Get individual objects
import GetArticle from './queries/GetArticle';
import GetReply from './queries/GetReply';
import GetUser from './queries/GetUser';
import GetCategory from './queries/GetCategory';
import ListArticles from './queries/ListArticles';
import ListReplies from './queries/ListReplies';
import ListCategories from './queries/ListCategories';
import ListArticleReplyFeedbacks from './queries/ListArticleReplyFeedbacks';

// Set individual objects
import CreateArticle from './mutations/CreateArticle';
import CreateReply from './mutations/CreateReply';
import CreateArticleReply from './mutations/CreateArticleReply';
import CreateCategory from './mutations/CreateCategory';
import CreateArticleCategory from './mutations/CreateArticleCategory';
import CreateOrUpdateArticleReplyFeedback from './mutations/CreateOrUpdateArticleReplyFeedback';
import CreateOrUpdateReplyRequestFeedback from './mutations/CreateOrUpdateReplyRequestFeedback';
import CreateOrUpdateArticleCategoryFeedback from './mutations/CreateOrUpdateArticleCategoryFeedback';
import CreateOrUpdateReplyRequest from './mutations/CreateOrUpdateReplyRequest';
import CreateOrUpdateUser from './mutations/CreateOrUpdateUser';
import UpdateArticleReplyStatus from './mutations/UpdateArticleReplyStatus';
import UpdateArticleCategoryStatus from './mutations/UpdateArticleCategoryStatus';
import UpdateUser from './mutations/UpdateUser';

export default new GraphQLSchema({
  query: new GraphQLObjectType({
    name: 'Query',
    fields: {
      GetArticle,
      GetReply,
      GetUser,
      GetCategory,
      ListArticles,
      ListReplies,
      ListCategories,
      ListArticleReplyFeedbacks,
    },
  }),
  mutation: new GraphQLObjectType({
    name: 'Mutation',
    fields: {
      CreateArticle,
      CreateReply,
      CreateArticleReply,
      CreateCategory,
      CreateArticleCategory,
      CreateReplyRequest: {
        ...CreateOrUpdateReplyRequest,
        deprecationReason: 'Use CreateOrUpdateReplyRequest instead',
      },
      CreateOrUpdateReplyRequest,
      CreateOrUpdateArticleReplyFeedback,
      CreateOrUpdateArticleCategoryFeedback,
      CreateOrUpdateReplyRequestFeedback,
      CreateOrUpdateUser,
      UpdateArticleReplyStatus,
      UpdateArticleCategoryStatus,
      UpdateUser,
    },
  }),
});
