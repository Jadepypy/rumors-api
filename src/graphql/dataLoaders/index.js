import docLoaderFactory from './docLoaderFactory';
import analyticsLoaderFactory from './analyticsLoaderFactory';
import articleRepliesByReplyIdLoaderFactory from './articleRepliesByReplyIdLoaderFactory';
import articleCategoriesByCategoryIdLoaderFactory from './articleCategoriesByCategoryIdLoaderFactory';
import articleReplyFeedbacksLoaderFactory from './articleReplyFeedbacksLoaderFactory';
import articleCategoryFeedbacksLoaderFactory from './articleCategoryFeedbacksLoaderFactory';
import searchResultLoaderFactory from './searchResultLoaderFactory';
import urlLoaderFactory from './urlLoaderFactory';
import repliedArticleCountLoaderFactory from './repliedArticleCountLoaderFactory';
import votedArticleReplyCountLoaderFactory from './votedArticleReplyCountLoaderFactory';
import userLevelLoaderFactory from './userLevelLoaderFactory';
import userLoaderFactory from './userLoaderFactory';

export default class DataLoaders {
  // List of data loaders
  //
  get docLoader() {
    return this._checkOrSetLoader('docLoader', docLoaderFactory);
  }
  get articleRepliesByReplyIdLoader() {
    return this._checkOrSetLoader(
      'articleRepliesByReplyIdLoader',
      articleRepliesByReplyIdLoaderFactory
    );
  }
  get articleCategoriesByCategoryIdLoader() {
    return this._checkOrSetLoader(
      'articleCategoriesByCategoryIdLoader',
      articleCategoriesByCategoryIdLoaderFactory
    );
  }
  get articleReplyFeedbacksLoader() {
    return this._checkOrSetLoader(
      'articleReplyFeedbacksLoader',
      articleReplyFeedbacksLoaderFactory
    );
  }
  get articleCategoryFeedbacksLoader() {
    return this._checkOrSetLoader(
      'articleCategoryFeedbacksLoader',
      articleCategoryFeedbacksLoaderFactory
    );
  }
  get searchResultLoader() {
    return this._checkOrSetLoader(
      'searchResultLoader',
      searchResultLoaderFactory
    );
  }

  get urlLoader() {
    return this._checkOrSetLoader('urlLoader', urlLoaderFactory);
  }

  get repliedArticleCountLoader() {
    return this._checkOrSetLoader(
      'repliedArticleCountLoader',
      repliedArticleCountLoaderFactory
    );
  }

  get votedArticleReplyCountLoader() {
    return this._checkOrSetLoader(
      'votedArticleReplyCountLoader',
      votedArticleReplyCountLoaderFactory
    );
  }

  get userLoader() {
    return this._checkOrSetLoader('userLoader', userLoaderFactory);
  }

  get userLevelLoader() {
    return this._checkOrSetLoader('userLevelLoader', userLevelLoaderFactory);
  }

  get analyticsLoader() {
    return this._checkOrSetLoader('analyticsLoader', analyticsLoaderFactory);
  }

  // inner-workings
  //
  constructor() {
    this._loaders = {};
  }

  _checkOrSetLoader(name, factoryFn) {
    if (this._loaders[name]) return this._loaders[name];

    this._loaders[name] = factoryFn(this);
    return this._loaders[name];
  }
}
