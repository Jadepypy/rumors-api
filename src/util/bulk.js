const BATCH_SIZE = 100;

export default class Bulk {
  constructor(client, batchSize = BATCH_SIZE) {
    this.client = client;
    this.batchSize = batchSize;
    this._operations = [];
    this.actionsCount = 0;
  }

  async push(items, count = 1) {
    this._operations.push(...items);
    this.actionsCount += count;
    if (
      this.actionCounts > this.batchSize ||
      this._operations.length > 3 * this.batchSize
    ) {
      await this.flush();
    }
    return this;
  }

  async flush() {
    if (this._operations.length === 0) {
      return this;
    }

    // Process entires queued up
    const op = this._operations.splice(0);
    this.actionCounts = 0;
    const { body } = await this.client.bulk({
      body: op,
      refresh: 'true',
    });
    if (body.errors) {
      console.error(body.errors);
    } else {
      console.info(`${body.items.length} items processed`);
    }

    return this;
  }
}
