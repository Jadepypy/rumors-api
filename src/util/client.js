import elasticsearch from '@elastic/elasticsearch';

export default new elasticsearch.Client({
  node: process.env.ELASTICSEARCH_URL,
});

// Processes {_id, _version, found, _source: {...}} to
// {id, ..._source}.
//
export function processMeta({
  _id: id,
  _source: source,

  found, // for mget queries
  _score, // for search queries

  sort, // cursor when sorted
}) {
  if (found || _score !== undefined) {
    return { id, ...source, _cursor: sort, _score };
  }
  return null; // not found
}
