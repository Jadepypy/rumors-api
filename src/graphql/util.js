import {
  GraphQLInputObjectType,
  GraphQLObjectType,
  GraphQLString,
  GraphQLInt,
  GraphQLList,
  GraphQLEnumType,
  GraphQLFloat,
  GraphQLNonNull,
  GraphQLID,
  GraphQLBoolean,
} from 'graphql';

import Connection from './interfaces/Connection';
import Edge from './interfaces/Edge';
import PageInfo from './interfaces/PageInfo';
import Highlights from './models/Highlights';
import client from 'util/client';
import { imageHash } from 'image-hash';

// https://www.graph.cool/docs/tutorials/designing-powerful-apis-with-graphql-query-parameters-aing7uech3
//
// Filtering args definition & parsing
//

/**
 * @param {string} typeName
 * @param {GraphQLScalarType} argType
 * @param {string} description
 * @returns {GraphQLInputObjectType}
 */
function getArithmeticExpressionType(typeName, argType, description) {
  return new GraphQLInputObjectType({
    name: typeName,
    description,
    fields: {
      LT: { type: argType },
      LTE: { type: argType },
      GT: { type: argType },
      GTE: { type: argType },
      EQ: { type: argType },
    },
  });
}

export const timeRangeInput = getArithmeticExpressionType(
  'TimeRangeInput',
  GraphQLString,
  'List only the entries that were created between the specific time range. ' +
    'The time range value is in elasticsearch date format (https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-date-format.html)'
);
export const intRangeInput = getArithmeticExpressionType(
  'RangeInput',
  GraphQLInt,
  'List only the entries whose field match the criteria.'
);

/**
 * @param {object} arithmeticFilterObj - {LT, LTE, GT, GTE, EQ}, the structure returned by getArithmeticExpressionType
 * @returns {object} Elasticsearch range filter param
 * @see https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-range-query.html#range-query-field-params
 */
export function getRangeFieldParamFromArithmeticExpression(
  arithmeticFilterObj
) {
  // EQ overrides all other operators
  if (typeof arithmeticFilterObj.EQ !== 'undefined') {
    return {
      gte: arithmeticFilterObj.EQ,
      lte: arithmeticFilterObj.EQ,
    };
  }

  const conditionEntries = Object.entries(arithmeticFilterObj);

  if (conditionEntries.length === 0) throw new Error('Invalid Expression!');

  return Object.fromEntries(
    conditionEntries.map(([key, value]) => [key.toLowerCase(), value])
  );
}

export const moreLikeThisInput = new GraphQLInputObjectType({
  name: 'MoreLikeThisInput',
  description:
    'Parameters for Elasticsearch more_like_this query.\n' +
    'See: https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-mlt-query.html',
  fields: {
    like: {
      type: GraphQLString,
      description: 'The text string to search for.',
    },
    minimumShouldMatch: {
      type: GraphQLString,
      description:
        'more_like_this query\'s "minimum_should_match" query param.\n' +
        'See https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-minimum-should-match.html for possible values.',
    },
  },
});

export function createFilterType(typeName, args) {
  const filterType = new GraphQLInputObjectType({
    name: typeName,
    fields: () => ({
      ...args,
      // TODO: converting nested AND / OR to elasticsearch
      // AND: { type: new GraphQLList(filterType) },
      // OR: { type: new GraphQLList(filterType) },
    }),
  });
  return filterType;
}

//
// Sort args definition & parsing
//

const SortOrderEnum = new GraphQLEnumType({
  name: 'SortOrderEnum',
  values: {
    ASC: { value: 'asc' },
    DESC: { value: 'desc' },
  },
});

/**
 * @param {string} typeName
 * @param {Array<string|{name: string, description: string}>} filterableFields
 * @returns {GraphQLList<GarphQLInputObjectType>} sort input type for an field input argument.
 */
export function createSortType(typeName, filterableFields = []) {
  return new GraphQLList(
    new GraphQLInputObjectType({
      name: typeName,
      description:
        'An entry of orderBy argument. Specifies field name and the sort order. Only one field name is allowd per entry.',
      fields: filterableFields.reduce((fields, field) => {
        const fieldName = typeof field === 'string' ? field : field.name;
        const description =
          typeof field === 'string' ? undefined : field.description;

        return {
          ...fields,
          [fieldName]: { type: SortOrderEnum, description },
        };
      }, {}),
    })
  );
}

export const pagingArgs = {
  first: {
    type: GraphQLInt,
    description: 'Returns only first <first> results',
    defaultValue: 10,
  },
  after: {
    type: GraphQLString,
    description:
      'Specify a cursor, returns results after this cursor. cannot be used with "before".',
  },
  before: {
    type: GraphQLString,
    description:
      'Specify a cursor, returns results before this cursor. cannot be used with "after".',
  },
};

/**
 * @param {object[]} orderBy - sort input object type
 * @param {{[string]: (order: object) => object}} fieldFnMap - Defines one elasticsearch sort argument entry for a field
 * @returns {Array<{[string]: {order: string}}>} Elasticsearch sort argument in query body
 */
export function getSortArgs(orderBy, fieldFnMap = {}) {
  return orderBy
    .map(item => {
      const field = Object.keys(item)[0];
      const order = item[field];
      const defaultFieldFn = o => ({ [field]: { order: o } });

      return (fieldFnMap[field] || defaultFieldFn)(order);
    })
    .concat({ _id: { order: 'desc' } }); // enforce at least 1 sort order for pagination
}

// sort: [{fieldName: {order: 'desc'}}, {fieldName2: {order: 'desc'}}, ...]
// This utility function reverts the direction of each sort params.
//
function reverseSortArgs(sort) {
  if (!sort) return undefined;
  return sort.map(item => {
    const field = Object.keys(item)[0];
    const order = item[field].order === 'desc' ? 'asc' : 'desc';
    return {
      [field]: {
        ...item[field],
        order,
      },
    };
  });
}

// Export for custom resolveEdges() and resolveLastCursor()
//
export function getCursor(cursor) {
  return Buffer.from(JSON.stringify(cursor)).toString('base64');
}

export function getSearchAfterFromCursor(cursor) {
  if (!cursor) return undefined;
  return JSON.parse(Buffer.from(cursor, 'base64').toString('utf8'));
}

async function defaultResolveTotalCount({
  first, // eslint-disable-line no-unused-vars
  before, // eslint-disable-line no-unused-vars
  after, // eslint-disable-line no-unused-vars
  ...searchContext
}) {
  return (await client.count({
    ...searchContext,
    body: {
      ...searchContext.body,

      // totalCount cannot support these
      sort: undefined,
      track_scores: undefined,
    },
  })).body.count;
}

export async function defaultResolveEdges(
  { first, before, after, ...searchContext },
  args,
  { loaders }
) {
  if (before && after) {
    throw new Error('Use of before & after is prohibited.');
  }

  const nodes = await loaders.searchResultLoader.load({
    ...searchContext,
    body: {
      ...searchContext.body,
      size: first,
      search_after: getSearchAfterFromCursor(before || after),

      // if "before" is given, reverse the sort order and later reverse back
      //
      sort: before
        ? reverseSortArgs(searchContext.body.sort)
        : searchContext.body.sort,
      highlight: {
        order: 'score',
        fields: {
          text: {
            number_of_fragments: 1, // Return only 1 piece highlight text
            fragment_size: 200, // word count of highlighted fragment
            type: 'plain',
          },
          reference: {
            number_of_fragments: 1, // Return only 1 piece highlight text
            fragment_size: 200, // word count of highlighted fragment
            type: 'plain',
          },
        },
        pre_tags: ['<HIGHLIGHT>'],
        post_tags: ['</HIGHLIGHT>'],
      },
    },
  });

  if (before) {
    nodes.reverse();
  }

  return nodes.map(
    ({ _score: score, highlight, inner_hits, _cursor, ...node }) => ({
      node,
      cursor: getCursor(_cursor),
      score,
      highlight,
      inner_hits,
    })
  );
}

async function defaultResolveLastCursor(
  {
    first, // eslint-disable-line no-unused-vars
    before, // eslint-disable-line no-unused-vars
    after, // eslint-disable-line no-unused-vars
    ...searchContext
  },
  args,
  { loaders }
) {
  const lastNode = (await loaders.searchResultLoader.load({
    ...searchContext,
    body: {
      ...searchContext.body,
      sort: reverseSortArgs(searchContext.body.sort),
    },
    size: 1,
  }))[0];

  return lastNode && getCursor(lastNode._cursor);
}

async function defaultResolveFirstCursor(
  {
    first, // eslint-disable-line no-unused-vars
    before, // eslint-disable-line no-unused-vars
    after, // eslint-disable-line no-unused-vars
    ...searchContext
  },
  args,
  { loaders }
) {
  const firstNode = (await loaders.searchResultLoader.load({
    ...searchContext,
    size: 1,
  }))[0];

  return firstNode && getCursor(firstNode._cursor);
}

async function defaultResolveHighlights(edge) {
  const { highlight: { text, reference } = {}, inner_hits } = edge;

  const hyperlinks = inner_hits?.hyperlinks.hits.hits?.map(
    ({
      _source: { url },
      highlight: {
        'hyperlinks.title': title,
        'hyperlinks.summary': summary,
      } = {},
    }) => ({
      url,
      title: title ? title[0] : undefined,
      summary: summary ? summary[0] : undefined,
    })
  );

  // Elasticsearch highlight returns an array because it can be multiple fragments,
  // We directly returns first element(text, title, summary) here because we set number_of_fragments to 1.
  return {
    text: text ? text[0] : undefined,
    reference: reference ? reference[0] : undefined,
    hyperlinks,
  };
}

// All search
//
export function createConnectionType(
  typeName,
  nodeType,
  {
    // Default resolvers
    resolveTotalCount = defaultResolveTotalCount,
    resolveEdges = defaultResolveEdges,
    resolveLastCursor = defaultResolveLastCursor,
    resolveFirstCursor = defaultResolveFirstCursor,
    resolveHighlights = defaultResolveHighlights,
  } = {}
) {
  return new GraphQLObjectType({
    name: typeName,
    interfaces: [Connection],
    fields: () => ({
      totalCount: {
        type: new GraphQLNonNull(GraphQLInt),
        description:
          'The total count of the entire collection, regardless of "before", "after".',
        resolve: resolveTotalCount,
      },
      edges: {
        type: new GraphQLNonNull(
          new GraphQLList(
            new GraphQLNonNull(
              new GraphQLObjectType({
                name: `${typeName}Edge`,
                interfaces: [Edge],
                fields: {
                  node: { type: new GraphQLNonNull(nodeType) },
                  cursor: { type: new GraphQLNonNull(GraphQLString) },
                  score: { type: GraphQLFloat },
                  highlight: {
                    type: Highlights,
                    resolve: resolveHighlights,
                  },
                },
              })
            )
          )
        ),
        resolve: resolveEdges,
      },
      pageInfo: {
        type: new GraphQLNonNull(
          new GraphQLObjectType({
            name: `${typeName}PageInfo`,
            interfaces: [PageInfo],
            fields: {
              lastCursor: {
                type: GraphQLString,
                resolve: resolveLastCursor,
              },
              firstCursor: {
                type: GraphQLString,
                resolve: resolveFirstCursor,
              },
            },
          })
        ),
        resolve: params => params,
      },
    }),
  });
}

/**
 * @param {{status: T}[]} entriesWithStatus - list of objects with "status" field
 * @param {T[]} statuses - list of status to keep
 * @returns {Object[]}
 */
export function filterByStatuses(entriesWithStatus, statuses) {
  return entriesWithStatus.filter(({ status }) => statuses.includes(status));
}

/**
 * @param {Buffer} fileBuffer
 * @param {ArticleTypeEnum} type The article type
 * @returns {string} The hash for identifying if two files are similar
 */
export async function getMediaFileHash(fileBuffer, type) {
  let hash = '';
  if (type === 'IMAGE') {
    hash = await new Promise((resolve, reject) => {
      imageHash({ data: fileBuffer }, 16, true, (error, data) => {
        if (error) {
          console.error(error);
          reject(error);
        } else {
          resolve(data);
        }
      });
    });
  }

  return hash;
}

export const DEFAULT_ARTICLE_REPLY_STATUSES = ['NORMAL'];
export const DEFAULT_ARTICLE_CATEGORY_STATUSES = ['NORMAL'];
export const DEFAULT_REPLY_REQUEST_STATUSES = ['NORMAL'];
export const DEFAULT_ARTICLE_REPLY_FEEDBACK_STATUSES = ['NORMAL'];

/**
 * @param {string} pluralEntityName - the name to display on argument description
 * @returns {object} GraphQL args for common list filters
 */
export function createCommonListFilter(pluralEntityName) {
  return {
    appId: {
      type: GraphQLString,
      description: `Show only ${pluralEntityName} created by a specific app.`,
    },
    userId: {
      type: GraphQLString,
      description: `Show only ${pluralEntityName} created by the specific user.`,
    },
    createdAt: {
      type: timeRangeInput,
      description: `List only the ${pluralEntityName} that were created between the specific time range.`,
    },
    ids: {
      type: new GraphQLList(new GraphQLNonNull(GraphQLID)),
      description: `If given, only list out ${pluralEntityName} with specific IDs`,
    },
    selfOnly: {
      type: GraphQLBoolean,
      description: `Only list the ${pluralEntityName} created by the currently logged in user`,
    },
  };
}

/**
 * Attach (mutates) filterQueries with Elasticsearch query objects by args.filter in GraphQL resolver
 *
 * @param {Array<Object>} filterQueries - list of filter queries of Elasticsearch bool query
 * @param {object} filter - args.filter in resolver
 * @param {string} userId - userId for the currently logged in user
 * @param {string} appid - appId for the currently logged in user
 */
export function attachCommonListFilter(filterQueries, filter, userId, appId) {
  ['userId', 'appId'].forEach(field => {
    if (!filter[field]) return;
    filterQueries.push({ term: { [field]: filter[field] } });
  });

  if (filter.createdAt) {
    filterQueries.push({
      range: {
        createdAt: getRangeFieldParamFromArithmeticExpression(filter.createdAt),
      },
    });
  }

  if (filter.ids) {
    filterQueries.push({ ids: { values: filter.ids } });
  }

  if (filter.selfOnly) {
    if (!userId) throw new Error('selfOnly can be set only after log in');
    filterQueries.push({ term: { userId } }, { term: { appId } });
  }
}
