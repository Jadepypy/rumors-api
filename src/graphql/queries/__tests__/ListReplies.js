import gql from 'util/GraphQL';
import { loadFixtures, unloadFixtures } from 'util/fixtures';
import { getCursor } from 'graphql/util';
import fixtures from '../__fixtures__/ListReplies';

describe('ListReplies', () => {
  beforeAll(() => loadFixtures(fixtures));

  it('lists all replies', async () => {
    expect(
      await gql`
        {
          ListReplies {
            totalCount
            edges {
              node {
                id
              }
              cursor
            }
            pageInfo {
              firstCursor
              lastCursor
            }
          }
        }
      `()
    ).toMatchSnapshot();
  });

  it('sorts', async () => {
    expect(
      await gql`
        {
          ListReplies(orderBy: [{ createdAt: DESC }]) {
            edges {
              node {
                id
              }
            }
            totalCount
            pageInfo {
              firstCursor
              lastCursor
            }
          }
        }
      `()
    ).toMatchSnapshot();
  });

  it('filters', async () => {
    expect(
      await gql`
        {
          ListReplies(
            filter: { moreLikeThis: { like: "foo", minimumShouldMatch: "5%" } }
          ) {
            edges {
              node {
                id
              }
            }
            totalCount
            pageInfo {
              firstCursor
              lastCursor
            }
          }
        }
      `()
    ).toMatchSnapshot();

    expect(
      await gql`
        {
          ListReplies(filter: { selfOnly: true }) {
            edges {
              node {
                id
              }
            }
            totalCount
            pageInfo {
              firstCursor
              lastCursor
            }
          }
        }
      `(
        {},
        {
          userId: 'foo',
          appId: 'test',
        }
      )
    ).toMatchSnapshot();

    expect(
      await gql`
        {
          ListReplies(filter: { type: RUMOR }) {
            edges {
              node {
                id
              }
            }
            totalCount
            pageInfo {
              firstCursor
              lastCursor
            }
          }
        }
      `()
    ).toMatchSnapshot();
  });

  it('filters by moreLikeThis and given text, find replies containing hyperlinks with the said text', async () => {
    expect(
      await gql`
        {
          ListReplies(
            filter: { moreLikeThis: { like: "「長鋏歸來乎！食無魚。」" } }
          ) {
            edges {
              node {
                id
              }
            }
          }
        }
      `()
    ).toMatchSnapshot();
  });

  it("filters by moreLikeThis and given text, find replies with the said URL's content", async () => {
    expect(
      await gql`
        {
          ListReplies(
            filter: {
              moreLikeThis: {
                like: "請看 http://foo.com"
                minimumShouldMatch: "5%"
              }
            }
          ) {
            edges {
              node {
                id
              }
            }
          }
        }
      `()
    ).toMatchSnapshot();
  });

  it('filters by time range', async () => {
    expect(
      await gql`
        {
          ListReplies(
            filter: { createdAt: { GT: "2020-02-06T00:00:00.000Z" } }
          ) {
            edges {
              node {
                id
              }
            }
            totalCount
          }
        }
      `()
    ).toMatchSnapshot();
    expect(
      await gql`
        {
          ListReplies(
            filter: { createdAt: { LTE: "2020-02-06T00:00:00.000Z" } }
          ) {
            edges {
              node {
                id
              }
            }
            totalCount
          }
        }
      `()
    ).toMatchSnapshot();
    expect(
      await gql`
        {
          ListReplies(
            filter: {
              createdAt: {
                GTE: "2020-02-04T00:00:00.000Z"
                LTE: "2020-02-06T00:00:00.000Z"
              }
            }
          ) {
            edges {
              node {
                id
              }
            }
            totalCount
          }
        }
      `()
    ).toMatchSnapshot();
  });

  it('filters by mixed query', async () => {
    // Mixes 'should' and 'filter' query. At least 1 'should' must match.
    // Therefore, this query should only match 2 results instead of all that satisfies type = NOT_ARTICLE

    expect(
      await gql`
        {
          ListReplies(
            filter: { type: NOT_ARTICLE, moreLikeThis: { like: "foo" } }
          ) {
            edges {
              node {
                id
              }
            }
            totalCount
          }
        }
      `()
    ).toMatchSnapshot();
  });

  it('supports after', async () => {
    expect(
      await gql`
        query($cursor: String) {
          ListReplies(after: $cursor) {
            edges {
              node {
                id
              }
            }
            totalCount
            pageInfo {
              firstCursor
              lastCursor
            }
          }
        }
      `({ cursor: getCursor(['moreLikeThis2']) })
    ).toMatchSnapshot();
  });

  it('supports before', async () => {
    expect(
      await gql`
        query($cursor: String) {
          ListReplies(before: $cursor) {
            edges {
              node {
                id
              }
            }
            totalCount
            pageInfo {
              firstCursor
              lastCursor
            }
          }
        }
      `({ cursor: getCursor(['moreLikeThis1']) })
    ).toMatchSnapshot();
  });

  it('handles selfOnly filter properly if not logged in', async () => {
    expect(
      await gql`
        {
          ListReplies(filter: { selfOnly: true }) {
            edges {
              node {
                id
              }
            }
            totalCount
            pageInfo {
              firstCursor
              lastCursor
            }
          }
        }
      `()
    ).toMatchSnapshot();
  });

  afterAll(() => unloadFixtures(fixtures));
});
