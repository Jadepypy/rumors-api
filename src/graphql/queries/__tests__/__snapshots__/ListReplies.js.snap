// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListReplies filters 1`] = `
Object {
  "data": Object {
    "ListReplies": Object {
      "edges": Array [
        Object {
          "node": Object {
            "id": "moreLikeThis2",
          },
        },
        Object {
          "node": Object {
            "id": "moreLikeThis1",
          },
        },
      ],
      "pageInfo": Object {
        "firstCursor": "WyJtb3JlTGlrZVRoaXMyIl0=",
        "lastCursor": "WyJtb3JlTGlrZVRoaXMxIl0=",
      },
      "totalCount": 2,
    },
  },
}
`;

exports[`ListReplies filters 2`] = `
Object {
  "data": Object {
    "ListReplies": Object {
      "edges": Array [
        Object {
          "node": Object {
            "id": "userFoo",
          },
        },
      ],
      "pageInfo": Object {
        "firstCursor": "WyJ1c2VyRm9vIl0=",
        "lastCursor": "WyJ1c2VyRm9vIl0=",
      },
      "totalCount": 1,
    },
  },
}
`;

exports[`ListReplies filters 3`] = `
Object {
  "data": Object {
    "ListReplies": Object {
      "edges": Array [
        Object {
          "node": Object {
            "id": "rumor",
          },
        },
      ],
      "pageInfo": Object {
        "firstCursor": "WyJydW1vciJd",
        "lastCursor": "WyJydW1vciJd",
      },
      "totalCount": 1,
    },
  },
}
`;

exports[`ListReplies filters by mixed query 1`] = `
Object {
  "data": Object {
    "ListReplies": Object {
      "edges": Array [
        Object {
          "node": Object {
            "id": "moreLikeThis2",
          },
        },
        Object {
          "node": Object {
            "id": "moreLikeThis1",
          },
        },
      ],
      "totalCount": 2,
    },
  },
}
`;

exports[`ListReplies filters by moreLikeThis and given text, find replies containing hyperlinks with the said text 1`] = `
Object {
  "data": Object {
    "ListReplies": Object {
      "edges": Array [
        Object {
          "node": Object {
            "id": "referenceUrl",
          },
        },
      ],
    },
  },
}
`;

exports[`ListReplies filters by moreLikeThis and given text, find replies with the said URL's content 1`] = `
Object {
  "data": Object {
    "ListReplies": Object {
      "edges": Array [
        Object {
          "node": Object {
            "id": "userFoo",
          },
        },
        Object {
          "node": Object {
            "id": "rumor",
          },
        },
      ],
    },
  },
}
`;

exports[`ListReplies filters by time range 1`] = `
Object {
  "data": Object {
    "ListReplies": Object {
      "edges": Array [
        Object {
          "node": Object {
            "id": "userFoo",
          },
        },
      ],
      "totalCount": 1,
    },
  },
}
`;

exports[`ListReplies filters by time range 2`] = `
Object {
  "data": Object {
    "ListReplies": Object {
      "edges": Array [
        Object {
          "node": Object {
            "id": "rumor",
          },
        },
        Object {
          "node": Object {
            "id": "referenceUrl",
          },
        },
        Object {
          "node": Object {
            "id": "moreLikeThis2",
          },
        },
        Object {
          "node": Object {
            "id": "moreLikeThis1",
          },
        },
      ],
      "totalCount": 4,
    },
  },
}
`;

exports[`ListReplies filters by time range 3`] = `
Object {
  "data": Object {
    "ListReplies": Object {
      "edges": Array [
        Object {
          "node": Object {
            "id": "rumor",
          },
        },
        Object {
          "node": Object {
            "id": "referenceUrl",
          },
        },
        Object {
          "node": Object {
            "id": "moreLikeThis2",
          },
        },
        Object {
          "node": Object {
            "id": "moreLikeThis1",
          },
        },
      ],
      "totalCount": 4,
    },
  },
}
`;

exports[`ListReplies handles selfOnly filter properly if not logged in 1`] = `
Object {
  "data": Object {
    "ListReplies": null,
  },
  "errors": Array [
    [GraphQLError: selfOnly can be set only after log in],
  ],
}
`;

exports[`ListReplies lists all replies 1`] = `
Object {
  "data": Object {
    "ListReplies": Object {
      "edges": Array [
        Object {
          "cursor": "WyJ1c2VyRm9vIl0=",
          "node": Object {
            "id": "userFoo",
          },
        },
        Object {
          "cursor": "WyJydW1vciJd",
          "node": Object {
            "id": "rumor",
          },
        },
        Object {
          "cursor": "WyJyZWZlcmVuY2VVcmwiXQ==",
          "node": Object {
            "id": "referenceUrl",
          },
        },
        Object {
          "cursor": "WyJtb3JlTGlrZVRoaXMyIl0=",
          "node": Object {
            "id": "moreLikeThis2",
          },
        },
        Object {
          "cursor": "WyJtb3JlTGlrZVRoaXMxIl0=",
          "node": Object {
            "id": "moreLikeThis1",
          },
        },
      ],
      "pageInfo": Object {
        "firstCursor": "WyJ1c2VyRm9vIl0=",
        "lastCursor": "WyJtb3JlTGlrZVRoaXMxIl0=",
      },
      "totalCount": 5,
    },
  },
}
`;

exports[`ListReplies sorts 1`] = `
Object {
  "data": Object {
    "ListReplies": Object {
      "edges": Array [
        Object {
          "node": Object {
            "id": "userFoo",
          },
        },
        Object {
          "node": Object {
            "id": "moreLikeThis1",
          },
        },
        Object {
          "node": Object {
            "id": "moreLikeThis2",
          },
        },
        Object {
          "node": Object {
            "id": "rumor",
          },
        },
        Object {
          "node": Object {
            "id": "referenceUrl",
          },
        },
      ],
      "pageInfo": Object {
        "firstCursor": "WzE1ODEwMzM2MDAwMDAsInVzZXJGb28iXQ==",
        "lastCursor": "WzE1ODA3NzQ0MDAwMDAsInJlZmVyZW5jZVVybCJd",
      },
      "totalCount": 5,
    },
  },
}
`;

exports[`ListReplies supports after 1`] = `
Object {
  "data": Object {
    "ListReplies": Object {
      "edges": Array [
        Object {
          "node": Object {
            "id": "moreLikeThis1",
          },
        },
      ],
      "pageInfo": Object {
        "firstCursor": "WyJ1c2VyRm9vIl0=",
        "lastCursor": "WyJtb3JlTGlrZVRoaXMxIl0=",
      },
      "totalCount": 5,
    },
  },
}
`;

exports[`ListReplies supports before 1`] = `
Object {
  "data": Object {
    "ListReplies": Object {
      "edges": Array [
        Object {
          "node": Object {
            "id": "userFoo",
          },
        },
        Object {
          "node": Object {
            "id": "rumor",
          },
        },
        Object {
          "node": Object {
            "id": "referenceUrl",
          },
        },
        Object {
          "node": Object {
            "id": "moreLikeThis2",
          },
        },
      ],
      "pageInfo": Object {
        "firstCursor": "WyJ1c2VyRm9vIl0=",
        "lastCursor": "WyJtb3JlTGlrZVRoaXMxIl0=",
      },
      "totalCount": 5,
    },
  },
}
`;
